meilisearch-core/src/bucket_sort.rs

```diff
@@ -331,10 +333,16 @@ where
                         let entry = key_cache.entry(document.id);
                         let key = entry.or_insert_with(|| (distinct)(document.id).map(Rc::new));
 
-                        match key.clone() {
+                        let distinct_accepted = match key.clone() {
```

Comment:
```suggestion
                for document in group.iter() {
                    let filter_accepted = match &filter {
                        Some(filter) => {
                            let entry = filter_map.entry(document.id);
                            let accepted = *entry.or_insert_with(|| (filter)(document.id));
                            if !accepted {
                                filtered_count += 1;
                            }
                            accepted
                        }
                        None => true,
                    };

```

I would suggest this instead, this is more straightforward I think

Reply:
@MarinPostma, I finally got around to this. Just for clarification, this diff should be applied in place of Lines 322 to 333 right? And I should just get rid of the `distinct_accepted` logic in the `if filter_accepted` block. However, when I tried this originally (and again now after your suggestion), for the `test search_with_filter`, I get 0 as the final count, presumably because it filters out everything in the sample set. When I tried this now, I got a `'attempt to subtract with overflow'` panic also. Am I missing something?

Reply:
You're right I have been a little too fast on this one, I have tried this one out and it seems to be working:

```rust
            for group in group.binary_group_by_mut(|a, b| criterion.eq(&ctx, a, b)) {
                // we must compute the real distinguished len of this sub-group
                for document in group.iter() {
                    let filter_accepted = match &filter {
                        Some(filter) => {
                            let entry = filter_map.entry(document.id);
                            *entry.or_insert_with(|| { 
                                let accepted = (filter)(document.id);
                                // we only want to count it out the first time we see it
                                if !accepted {
                                    filtered_count += 1;
                                }
                                accepted
                            })
                        }
                        None => true,
                    };

                    if filter_accepted {
                        let entry = key_cache.entry(document.id);
                        let mut seen = true;
                        let key = entry.or_insert_with(|| {
                            seen = false;
                            (distinct)(document.id).map(Rc::new)
                        });

                        let distinct = match key.clone() {
                            Some(key) => buf_distinct.register(key),
                            None => buf_distinct.register_without_key(),
                        };

                        // we only want to count the document if it is the first time we see it and
                        // if it wasn't accepted by distinct
                        if !seen && !distinct {
                            filtered_count += 1;
                        }
                    } 
                    // the requested range end is reached: stop computing distinct
                    if buf_distinct.len() >= range.end {
                        break;
                    }
                }

                documents_seen += group.len();
                groups.push(group);

                // if this sub-group does not overlap with the requested range
                // we must update the distinct map and its start index
                if buf_distinct.len() < range.start {
                    buf_distinct.transfert_to_internal();
                    distinct_raw_offset = documents_seen;
                }

                // we have sort enough documents if the last document sorted is after
                // the end of the requested range, we can continue to the next criterion
                if buf_distinct.len() >= range.end {
                    continue 'criteria;
                }
            }
```

these are the tests I have been trying it on:
```rust
#[actix_rt::test]
async fn test_filter_nb_hits_search_normal() {
    let mut server = common::Server::with_uid("test");

    let body = json!({
        "uid": "test",
        "primaryKey": "id",
    });

    server.create_index(body).await;
    let documents = json!([
        {
            "id": 1,
            "content": "a",
            "color": "green",
            "size": 1,
        },
        {
            "id": 2,
            "content": "a",
            "color": "green",
            "size": 2,
        },
        {
            "id": 3,
            "content": "a",
            "color": "blue",
            "size": 3,
        },
    ]);

    server.add_or_update_multiple_documents(documents).await;
    let (response, _) = server.search_post(json!({"q": "a"})).await;
    assert_eq!(response["nbHits"], 3);

    let (response, _) = server.search_post(json!({"q": "a", "filters": "size = 1"})).await;
    assert_eq!(response["nbHits"], 1);

    server.update_distinct_attribute(json!("color")).await;

    let (response, _) = server.search_post(json!({"q": "a"})).await;
    assert_eq!(response["nbHits"], 2);

    let (response, _) = server.search_post(json!({"q": "a", "filters": "size < 3"})).await;
    println!("result: {}", response);
    assert_eq!(response["nbHits"], 1);
}
```

Reply:
the reason for the subtracting with overflow is that we are counting some items more than once (as you've probably figured), we really want to count them only once we first see them :)

Reply:
we basically have to try the same tests on placeholder search now